lib/kcl.js

```
@@ -153,41 +154,73 @@ module.exports = function(config, kinesis) {
-          throttledPutToCloudwatch('ShardIteratorAgeInMs', (+new Date()) - resp.Records[0].ApproximateArrivalTimestamp, 'Milliseconds', shard.id);
-        }
+    var req = kinesis.getRecords({ShardIterator: shard.iterator, Limit: config.limit});
+    var requestTimeOut = setTimeout(req.abort.bind(req), 5000);
 
-        // malformed responses will have an undefined NextShardIterator. In that case, we keep the existing one.
-        shard.iterator = resp.NextShardIterator === undefined ? shard.iterator : resp.NextShardIterator;
-        shard.lastGetRecords = +new Date();
+    req.on('error', function (err) {
+      if (err.code === 'SyntaxError' || err.code === 'RequestAbortedError') {
+        if (config.verbose) logger.info('[ getRecords ] Found SyntaxError', err);
+        return setTimeout(getRecords.bind(this, shard), 500);
+      }
+      if (err.code === 'ProvisionedThroughputExceededException') {
+        if (config.verbose) logger.info('[ getRecords ] Found ProvisionedThroughputExceededException', err);
+        return setTimeout(getRecords.bind(this, shard), 1000);
+      }
+      if (err.code === 'ServiceUnavailable' || err.code === 'InternalFailure' || err.code === 'NetworkingError') {
+        if (config.verbose) logger.info('[ getRecords ] Found serviceUnavailable', err);
+        return setTimeout(getRecords.bind(this, shard), 2000);
+      }
+      throw new Error(err);
+    });
+    req.on('extractData', function (response) {
+      if (config.verbose) logger.info('[ getRecords ] extractData', response.data.Records.length, 'items');
+    });
+    req.on('retry', function (resp) {
+      if (config.verbose) logger.info('[ getRecords retry ] retry count: ', resp.retryCount);
+      var nonRetryableErrors = ['ServiceUnavailable', 'InternalFailure', 'NetworkingError', 'ProvisionedThroughputExceededException', 'SyntaxError', 'RequestAbortedError'];
+      if (nonRetryableErrors.indexOf(resp.error.code) >= 0) {
```

Yeah I'm not sure either. I've gone back and made another dev build based off of master with:

- log `err, resp` immediately in the `kinesis.getRecords()` callback
- 1 `maxRetries`,  5 seconds `connectTimeout`, 5 seconds `timeout`

And yet, there is absolutely nothing logged (neither err, nor resp) before a "max processing time reached" error, which tells me that the request has gone wrong, and something went wrong with retries _or_ it never actually retries and gets stuck in limbo. It should throw an error one way or another.

These low retries/timeouts seem to also yield a lot more `ProvisionedThroughputExceededException` errors